x/likenft/keeper/msg_server_new_class.go

```diff
@@ -14,44 +15,77 @@ import (
+	var safeParent types.ClassParent
+	var safeUserAddress sdk.AccAddress
+	if msg.Parent.Type == types.ClassParentType_ISCN {
+		// Assert iscn id is valid
+		iscnId, err := iscntypes.ParseIscnId(msg.Parent.IscnIdPrefix)
+		if err != nil {
+			return nil, types.ErrInvalidIscnId.Wrapf("%s", err.Error())
+		}
+		// Assert iscn exists
+		iscnRecord := k.iscnKeeper.GetContentIdRecord(ctx, iscnId.Prefix)
+		if iscnRecord == nil {
+			return nil, types.ErrIscnRecordNotFound.Wrapf("ISCN %s not found", iscnId.Prefix.String())
+		}
+		// Assert current user is owner
+		safeUserAddress, err = sdk.AccAddressFromBech32(msg.Creator)
+		if err != nil {
+			return nil, sdkerrors.ErrInvalidAddress.Wrapf("%s", err.Error())
+		}
+		if !iscnRecord.OwnerAddress().Equals(safeUserAddress) {
+			return nil, sdkerrors.ErrUnauthorized.Wrapf("%s is not the owner of the ISCN %s", msg.Creator, iscnId.Prefix.String())
+		}
+		safeParent = types.ClassParent{
+			Type:              types.ClassParentType_ISCN,
+			IscnIdPrefix:      iscnId.Prefix.String(),
+			IscnVersionAtMint: iscnRecord.LatestVersion,
+		}
+	} else if msg.Parent.Type == types.ClassParentType_ACCOUNT {
+		var err error
+		safeUserAddress, err = sdk.AccAddressFromBech32(msg.Creator)
+		if err != nil {
+			return nil, sdkerrors.ErrInvalidAddress.Wrapf("%s", err.Error())
+		}
+		safeParent = types.ClassParent{
+			Type:    types.ClassParentType_ACCOUNT,
+			Account: safeUserAddress.String(),
+		}
+	} else {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("Unsupported parent type %s", msg.Parent.Type.String())
 	}
 
 	// Make class id
 	var existingClassIds []string
-	value, found := k.GetClassesByISCN(ctx, iscnId.Prefix.String())
-	if found {
-		existingClassIds = value.ClassIds
-	}
-	newClassId, err := types.NewClassId(iscnId.Prefix.String(), len(existingClassIds))
-	if newClassId == nil || err != nil {
-		return nil, types.ErrFailedToMarshalData.Wrapf("%s", err.Error())
+	var newClassId string
+	if safeParent.Type == types.ClassParentType_ISCN {
+		value, found := k.GetClassesByISCN(ctx, safeParent.IscnIdPrefix)
+		if found {
+			existingClassIds = value.ClassIds
+		}
+		var err error
+		newClassId, err = types.NewClassIdForISCN(safeParent.IscnIdPrefix, len(existingClassIds))
+		if newClassId == "" || err != nil {
+			return nil, types.ErrFailedToMarshalData.Wrapf("%s", err.Error())
+		}
+	} else if safeParent.Type == types.ClassParentType_ACCOUNT {
+		value, found := k.GetClassesByAccount(ctx, safeUserAddress)
+		if found {
+			existingClassIds = value.ClassIds
+		}
+		var err error
+		newClassId, err = types.NewClassIdForAccount(safeUserAddress, len(existingClassIds))
+		if newClassId == "" || err != nil {
+			return nil, types.ErrFailedToMarshalData.Wrapf("%s", err.Error())
+		}
+	} else {
+		panic(fmt.Sprintf("Unsupported parent type %s after initial check", safeParent.Type.String()))
 	}
```

Comment:
I think this function is a bit too huge, consider creating helper functions and using switch cases? since this and the other PRs are very large perhaps we can handle this in another issue.

Reply:
https://github.com/oursky/likecoin-chain/issues/178